### PR TITLE
add ff-analogy

### PR DESCRIPTION
**Pull request by NianhengWu**
Take a test file contains 3 words a line separated by a space. If the words are A B C, then it should return words that qualify as D in A is to B as C is to D.

mostly the same as ff-similarity. Except it takes 3 words instead of 1, and invoke analogy function.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -197,13 +197,12 @@ dependencies = [
 [[package]]
 name = "finalfusion"
 version = "0.5.0"
+source = "git+https://github.com/finalfusion/finalfusion-rust.git#b5a0b1f3d8dda9d2b5abc13710cd79f54ff29ce4"
```
Comment (danieldk) [inline Cargo.lock]: Changes to `Cargo.lock` should not be part of the pull request. It is unrelated to the addition of `ff-anology`.

File: src/bin/ff-analogy.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+];
+
+fn parse_args() -> ArgMatches<'static> {
+    App::new("ff-analogy")
+        .settings(DEFAULT_CLAP_SETTINGS)
+        .arg(
+            Arg::with_name("format")
+                .short("f")
+                .value_name("FORMAT")
+                .help("Embedding format: finalfusion, finalfusion_mmap, word2vec, text, or textdims (default: finalfusion)")
```
Comment (danieldk) [inline src/bin/ff-analogy.rs]: This is not yet done in `ff-similar`, but the possible options can be specified with `possible_values`:

https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.possible_values

The default value with `default_value`:

https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.default_value

File: src/bin/ff-analogy.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("neighbors")
+                .short("k")
+                .value_name("K")
+                .help("Return K nearest neighbors (default: 10)")
```
Comment (danieldk) [inline src/bin/ff-analogy.rs]: This can also use `default_value`.

File: src/bin/ff-analogy.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    k: usize,
+}
+
+fn config_from_matches<'a>(matches: &ArgMatches<'a>) -> Config {
+    let embeddings_filename = matches.value_of("EMBEDDINGS").unwrap().to_owned();
+
+    let embedding_format = matches
+        .value_of("format")
+        .map(|f| EmbeddingFormat::try_from(f).or_exit("Cannot parse embedding format", 1))
+        .unwrap_or(EmbeddingFormat::FinalFusion);
```
Comment (danieldk) [inline src/bin/ff-analogy.rs]: This becomes `unwrap` when `default_value` is used above. The same for unwrapping `k` below.

File: src/bin/ff-analogy.rs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    let input = Input::from(matches.value_of("INPUT"));
+    let reader = input.buf_read().or_exit("Cannot open input for reading", 1); 
+
+    for line in reader.lines() {
+        let line = line.or_exit("Cannot read line", 1).trim().to_owned();
+        if line.is_empty() {
+            continue;
+        }
+
+        if line.starts_with(": ") {
```
Comment (danieldk) [inline src/bin/ff-analogy.rs]: This should go, since it breaks if someone would want to do a punctuation analogy ;). No hidden behavior (these sections are only used in the typical word2vec evaluation files, but we have `ff-compute-accuracy` for that).

File: src/bin/ff-analogy.rs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            println!("Section {}", line);
+            continue;
+        }
+
+        let split_line: Vec<&str> = line.split_whitespace().collect();
+
```
Comment (danieldk) [inline src/bin/ff-analogy.rs]: Superfluous whitespace. In general, use one empty line between.

File: src/bin/ff-analogy.rs (hunk 5 above)
Comment (NianhengWu) [reply src/bin/ff-analogy.rs]: Ah okay, so here the assumption of the input file is just like:
word1.1 word1.2 word1.3
word2.1 word2.2 word2.3
...

**Comment by NianhengWu** [comment]
> Great first PR!
> 
> I have added some comments to the PR, could you make the requested changes?
> 
> Also, for the PR to be merged, it needs to be squashed into a single commit. See: https://github.com/sfb833-a3/wiki/wiki/Pull-requests#before-accepting-a-pr

Will do, thanks for the feedback!

**Comment by danieldk** [comment]
This change now adds both `ff-analogy` and does the changes to the other files in one pr/commit. Separate changes should be in separate PRs and commits. I have split the commits and will open new PRs for both.
